src/input/inputModule.py

```diff
@@ -11,8 +13,34 @@ def all_symbol_names(self) -> list[str]:
         """Return every symbol that is imported when importing * from the library"""
         return []
 
-    def visit_call(self, func_name: str, args: list, kwargs: dict) -> IRNode:
-        pass
+    def visit_method(self, func_name: str, args: list, kwargs: dict) -> Optional[IRNode]:
+        visitor_method_name = f"visit_{func_name}"
 
-    def visit_call_on_ir_node(self, ir_node: IRNode, func_name: str, args: list, kwargs: dict) -> IRNode:
-        pass
+        if not hasattr(self, visitor_method_name):
+            print(f'Warning: Call of method "{func_name}" of library "{self.module_name}" has no handler.')
+            return None
+
+        visitor_method = getattr(self, visitor_method_name)
+        result = visitor_method(*args, **kwargs)
+        if not result:
+            print(f'Warning: Call of method "{func_name}" of library "{self.module_name}" returns no IR.')
+            return None
+        if not result.library:
+            result.library = self.module_name
+        return result
+
+    def visit_df_method(self, ir_node: IRNode, func_name: str, args: list, kwargs: dict) -> Optional[IRNode]:
+        visitor_method_name = f"visit_df_{func_name}"
+
+        if not hasattr(self, visitor_method_name):
+            print(f'Warning: Call of method "{func_name}" on dataframe of library "{self.module_name}" has no handler.')
+            return None
+
+        visitor_method = getattr(self, visitor_method_name)
+        result = visitor_method(ir_node, *args, **kwargs)
+        if not result:
+            print(f'Warning: Call of method "{func_name}" on dataframe of library "{self.module_name}" returns no IR.')
+            return None
```

Comment:
you should do one of both:
1. you raise an error
2. you return an empty node
since we know that None is not a good pattern

Reply:
that would remove the Optional from the return type. That would also mean that lines 46-50 of the NodeSelector would try to return the result of "method" (which will not be allowed to be None) and hanfle the exceptions (which in our case will occur most of the time, e.g. for non library code) by returning the cst_node. What do you say?

Reply:
Okay 👍